### PR TITLE
2533 Improve performance of Time POCO

### PR DESCRIPTION
**Pull request by ewoutkramer**
Added use of a cached value to speed up processing.

### CODE REVIEW

File: src/Hl7.Fhir.Base/Model/Time.cs [hunk 1]
```diff
@@ -30,15 +30,106 @@
...
+        public const string FMT_FULL = "HH:mm:ss";
+
+        public const string FMT_HOURMINSEC = "{0:D2}:{1:D2}:{2:D2}";
+
+        public Time(int hour, int minute, int second) : this(string.Format(CultureInfo.InvariantCulture, FMT_HOURMINSEC, hour, minute, second))
+        {
+            // Nothing
+        }
+
+        public static Time FromDateTimeOffset(DateTimeOffset dto) => new(dto.ToString(FMT_FULL));
```
Comment (marcovisserFurore) [inline src/Hl7.Fhir.Base/Model/Time.cs:54]: Can we not do it like this:

```
public static Time FromDateTimeOffset(DateTimeOffset dto) => new(dto.Hour, dto.Minute, dto.Second);
```
Then you do not need the `FMT_FULL`..

File: src/Hl7.Fhir.Base/Model/Time.cs [hunk 2]
```diff
@@ -30,15 +30,106 @@
...
+        public P.Time? ToTime() => TryToTime(out var dt) ? dt : throw new FormatException($"String '{Value}' was not recognized as a valid time.");
+
+        protected override void OnObjectValueChanged()
+        {
+            _parsedValue = null;
+            base.OnObjectValueChanged();
+        }
+
+        /// <summary>
+        /// Converts this Fhir Fhir Time to a <see cref="TimeSpan"/>.
```
Comment (marcovisserFurore) [inline src/Hl7.Fhir.Base/Model/Time.cs:105]:
```suggestion
        /// Converts this Fhir Time to a <see cref="TimeSpan"/>.
```

**Review by marcovisserFurore — CHANGES_REQUESTED** [review]
See some remarks/questions

File: src/Hl7.Fhir.Base/Model/Time.cs [hunk 3]
```diff
@@ -30,15 +30,106 @@
...
 #nullable enable
 
+using System;
+using System.Globalization;
+using P = Hl7.Fhir.ElementModel.Types;
+
 namespace Hl7.Fhir.Model
 {
 
     public partial class Time
```
[the remark this answers is not in the document]
Comment (ewoutkramer) [reply src/Hl7.Fhir.Base/Model/Time.cs:40]: Indeed. So I used this to align the datatypes better too. A bit of yak shaving of course....

File: src/Hl7.Fhir.Base/ElementModel/Types/Time.cs [hunk 4]
```diff
@@ -31,47 +32,57 @@ private Time(string original, DateTimeOffset parsedValue, DateTimePrecision prec
...
+        /// Converts the time to a full DateTimeOffset instance.
+        /// </summary>
+        /// <param name="year">Year used to turn a time into a date</param>
+        /// <param name="month">Month used to turn a time into a date</param>
+        /// <param name="day">Day used to turn a time into a date</param>
+        /// <param name="defaultOffset">Offset used when the time does not specify one.</param>
+        /// <returns></returns>
+        public DateTimeOffset ToDateTimeOffset(int year, int month, int day, TimeSpan defaultOffset)
+        {
+            var ticks = _value.Ticks % 10000;
```
[the remark this answers is not in the document]
Comment (ewoutkramer) [reply src/Hl7.Fhir.Base/ElementModel/Types/Time.cs:72]: Right, this requires more explanation.

**Review by marcovisserFurore — APPROVED** [review]

**Pull request merged**